### PR TITLE
Add calls to specialized versions of scan/verify intrinsics. 

### PR DESCRIPTION
**Pull request by mleair**
Add runtime calls to specialized versions of scan/verify intrinsics. Refactor some code into functions that operate on both scan/verify. Update the "adding intrinsics" tutorial.

### CODE REVIEW

File: flang/include/flang/Lower/CharacterRuntime.h [hunk 1]
```diff
@@ -66,11 +59,20 @@ void genTrim(Fortran::lower::FirOpBuilder &builder, mlir::Location loc,
              mlir::Value resultBox, mlir::Value stringBox);
 
 /// Generate call to verify runtime.
-/// This calls the descriptor based runtime call implementation of the verify
-/// intrinsic.
-void genVerify(Fortran::lower::FirOpBuilder &builder, mlir::Location loc,
+/// This calls the descriptor based runtime call implementation of the scan or
+/// verify intrinsicis.
```
Comment (klausler) [inline flang/include/flang/Lower/CharacterRuntime.h:63]: intrinsics*

File: flang/include/flang/Lower/CharacterRuntime.h [hunk 2]
```diff
@@ -66,11 +59,20 @@ void genTrim(Fortran::lower::FirOpBuilder &builder, mlir::Location loc,
...
+void genScanVerify(Fortran::lower::FirOpBuilder &builder, mlir::Location loc,
```
Comment (klausler) [inline flang/include/flang/Lower/CharacterRuntime.h:64]: INDEX is also related

File: flang/include/flang/Lower/CharacterRuntime.h [hunk 3]
```diff
@@ -66,11 +59,20 @@ void genTrim(Fortran::lower::FirOpBuilder &builder, mlir::Location loc,
              mlir::Value resultBox, mlir::Value stringBox);
 
 /// Generate call to verify runtime.
-/// This calls the descriptor based runtime call implementation of the verify
-/// intrinsic.
-void genVerify(Fortran::lower::FirOpBuilder &builder, mlir::Location loc,
+/// This calls the descriptor based runtime call implementation of the scan or
+/// verify intrinsics.
+void genScanVerify(Fortran::lower::FirOpBuilder &builder, mlir::Location loc,
```
Comment (schweitzpgi) [inline flang/include/flang/Lower/CharacterRuntime.h:64]: Please split this into two entry points: genScanDescriptor and genVerifyDescriptor to be consistent with genIndexDescriptor.

File: flang/include/flang/Lower/CharacterRuntime.h [hunk 4]
```diff
@@ -66,11 +59,20 @@ void genTrim(Fortran::lower::FirOpBuilder &builder, mlir::Location loc,
...
                mlir::Value resultBox, mlir::Value stringBox,
-               mlir::Value setBox, mlir::Value backBox, mlir::Value kind);
+               mlir::Value setBox, mlir::Value backBox, mlir::Value kind,
+               bool isScan);
+
+/// Generate call to the scan or verify runtime routine that is specialized on 
+/// \param kind.
+/// The \param kind represents the kind of the elements in the strings.
+mlir::Value genScanVerifyKind(Fortran::lower::FirOpBuilder &builder,
```
Comment (schweitzpgi) [inline flang/include/flang/Lower/CharacterRuntime.h:72]: Please split this into genScan and genVerify to be consistent with genIndex.

File: flang/lib/Lower/CharacterRuntime.cpp [hunk 5]
```diff
@@ -157,15 +157,19 @@ void Fortran::lower::genTrim(Fortran::lower::FirOpBuilder &builder,
...
-  auto scanFunc = getRuntimeFunc<mkRTKey(Scan)>(loc, builder);
-  auto fTy = scanFunc.getType();
+Fortran::lower::genScanVerify(Fortran::lower::FirOpBuilder &builder,
+                              mlir::Location loc, mlir::Value resultBox,
+                              mlir::Value stringBox, mlir::Value setBox,
+                              mlir::Value backBox, mlir::Value kind, 
+                              bool isScan) {
+
+  auto func = isScan ? getRuntimeFunc<mkRTKey(Scan)>(loc, builder)
+                     : getRuntimeFunc<mkRTKey(Verify)>(loc, builder);
```
Comment (schweitzpgi) [inline flang/lib/Lower/CharacterRuntime.cpp:171]: The different entry points can all share a common implementation that takes the specific getRuntimeFunc as an argument.

File: flang/lib/Lower/IntrinsicCall.cpp [hunk 6]
```diff
@@ -1693,35 +1659,85 @@ static mlir::Value createExtremumCompare(mlir::Location loc,
...
+                                llvm::ArrayRef<fir::ExtendedValue> args) {
+
+  return IntrinsicLibrary::genScanVerify(resultType, args, false);
+}
+
+// Process Scan/Verify intrinsic
+fir::ExtendedValue
+IntrinsicLibrary::genScanVerify(mlir::Type resultType,
+                                llvm::ArrayRef<fir::ExtendedValue> args,
+                                bool isScan) {
```
Comment (schweitzpgi) [inline flang/lib/Lower/IntrinsicCall.cpp:1671]: templates might be a more consistent option for code sharing instead of an extra `bool` argument.

**Review by schweitzpgi — COMMENTED** [review]
We tend to use templates for code sharing in flang. Do you want to take a stab at that?

File: flang/lib/Lower/CharacterRuntime.cpp [hunk 7]
```diff
@@ -157,54 +155,126 @@ void Fortran::lower::genTrim(Fortran::lower::FirOpBuilder &builder,
...
-  auto scanFunc = getRuntimeFunc<mkRTKey(Scan)>(loc, builder);
-  auto fTy = scanFunc.getType();
-  auto sourceFile = Fortran::lower::locationToFilename(builder, loc);
-  auto sourceLine =
-       Fortran::lower::locationToLineNo(builder, loc, fTy.getInput(6));
+Fortran::lower::genScanDescriptor(Fortran::lower::FirOpBuilder &builder,
+                                  mlir::Location loc, mlir::Value resultBox,
+                                  mlir::Value stringBox, mlir::Value setBox,
+                                  mlir::Value backBox, mlir::Value kind) {
+
```
Comment (schweitzpgi) [inline flang/lib/Lower/CharacterRuntime.cpp:166]: nit: blank line

File: flang/lib/Lower/CharacterRuntime.cpp [hunk 8]
```diff
@@ -157,54 +155,126 @@ void Fortran::lower::genTrim(Fortran::lower::FirOpBuilder &builder,
...
-/// TODO: Add specialized kind versions.
+/// Generate call to verify runtime routine.
+/// This calls the descriptor based runtime call implementation of the 
+/// verify intrinsic.
 void
+Fortran::lower::genVerifyDescriptor(Fortran::lower::FirOpBuilder &builder,
+                                    mlir::Location loc, mlir::Value resultBox,
+                                    mlir::Value stringBox, mlir::Value setBox,
+                                    mlir::Value backBox, mlir::Value kind) {
+
```
Comment (schweitzpgi) [inline flang/lib/Lower/CharacterRuntime.cpp:214]: nit: blank line

File: flang/lib/Lower/IntrinsicCall.cpp [hunk 9]
```diff
@@ -12,6 +12,7 @@
 // module.
 //
 //===----------------------------------------------------------------------===//
+#include "../../runtime/character.h"
```
Comment (schweitzpgi) [inline flang/lib/Lower/IntrinsicCall.cpp:15]: Is this required?

**Review by schweitzpgi — COMMENTED** [review]
Looks good.

File: flang/lib/Lower/IntrinsicCall.cpp (hunk 9 above)
Comment (mleair) [reply flang/lib/Lower/IntrinsicCall.cpp]: I  just removed it. :)

**Comment by sscalpone** [comment]
@mleair  Where are the tests?

**Comment by mleair** [comment]
> @mleair Where are the tests?

@sscalpone The tests are in flang/test/Lower/intrinsic-procedures.f90.

**Pull request merged**